_posts/2018-05-02-load-balanced-smart-proxies-with-haproxy.md

```diff
@@ -0,0 +1,138 @@
+
+### Yum metadata
+
+Each smart proxy generates its own yum metadata for RPM repositories. This means that if a request to `https://smartproxy.example.test/repo/repodata/repomd.xml` goes to the first smart proxy, and the next request to fetch `repodata/<CHECKSUM>-primary.xml.gz` goes to the second smart proxy, the second request will 404. To work around this, we advise that port 443 should be sticky-sessioned.
+
+## Requirements
+* a working Foreman/Katello installation (we don't need to do any changes here)
+* two or more machines for the Smart Proxy cluster
+* a load balancer - you can use a machine with HAProxy if you don't already have a load balancer
+* a common name in the DNS that points to the load balancer (we'll be using `smartproxy.example.test`)
+
+## Installing the first Smart Proxy
+### Generating Katello certificates
+When calling `foreman-proxy-certs-generate`, we have to pass `--foreman-proxy-cname smartproxy.example.test` to add the load-balanced name as an accepted `subjectAltName` to the list of names in the generated certificates. This allows clients to connect to each member of the smartproxy cluster as `smartproxy.example.test`
+
+### Preparing `custom-hiera.yaml`
+We need to configure Pulp to serve redirects in the streamer code to the name of the load-balancer, but this option is not exposed in the Installer. Thus edit `/etc/foreman-installer/custom-hiera.yaml` and add the following line at the end:
+
+```yaml
+pulp::lazy_redirect_host: smartproxy.example.test
+```
+### Running the installer
+As only the first Smart Proxy will host a PuppetCA, we configure that explicitly during the installation and also configure the Puppet certificate to include the `smartproxy.example.test` name:
+```
+--puppet-dns-alt-names smartproxy.example.test
+--puppet-ca-server proxy01.example.test
+--foreman-proxy-puppetca true
+--puppet-server-ca true
+```
+
+## Installing the other Smart Proxies
+The installation of all other Smart Proxies is almost identical to the first one. We just have to prepare the Puppet certs on the first one (as it's the only one having a PuppetCA) and disable the PuppetCA in the installer.
+
+### Generating Katello certificates
+This step is identical to the steps used to generate the certificates for the first smart proxy. Simply use `foreman-proxy-certs-generate` as described above.
+
+### Generating Puppet certificates
+On the first smart proxy, generate the Puppet certificates by running
+
+```shell
+puppet cert generate proxy02.example.test --dns_alt_names=smartproxy.example.test
+```
+
+Now copy `/etc/puppetlabs/puppet/ssl/*/proxy02.example.test.pem` and `/etc/puppetlabs/puppet/ssl/certs/ca.pem` to the second smart proxy and ensure the permissions are identical to the ones on the first host. The `puppet` user is created after installing the `puppetserver` package, so you might want to do that first. Also don't forget to check the SELinux contexts of the files you copied over (hint: `ls -lZ`) :)
```

Comment:
Should proxy02 point to proxy01 as a CA and have its own CA functionality disabled? I think you can also put the alt name in puppet.conf and then just request as normal. It does mean you need to sign the cert on proxy01, but then you don't have to transfer files and worry about permissions. Probably also works better with renewals.